File: sdk/runanywhere-python/runanywhere/inputs.py
"""Input value types shared by the namespaces (audio, image, messages, model refs)."""

from __future__ import annotations

import os
import struct
import tempfile
from dataclasses import dataclass, field
from enum import IntEnum
from typing import Any, Dict, List, Optional

import numpy as np

from ._generated_defaults import AudioCaptureDefaults
from .audio import decode_wav, downsample, float32_to_pcm16, pcm16_bytes
from .errors import SDKException

__all__ = [
    "AudioEncoding",
    "AudioFormat",
    "AudioFormatSpec",
    "AudioInput",
    "ChatMessage",
    "ImageInput",
    "JsonSchema",
    "ModelCategory",
    "ModelRef",
    "ModelRegistration",
    "ModelFilter",
    "Role",
    "STT_SAMPLE_RATE",
    "ToolDefinition",
    "RagDocument",
]

#: A JSON Schema, as a plain dict (what ``llm.generate_structured`` constrains decoding to).
JsonSchema = Dict[str, Any]

# The bridge transcribes mono PCM16 at the capture rate only, so every AudioInput
# normalizes to that. The rate itself is centralized in idl/sdk_defaults.proto.
STT_SAMPLE_RATE = AudioCaptureDefaults.MIC_SAMPLE_RATE_HZ


class AudioEncoding(IntEnum):
    """How the bytes of an :class:`AudioInput` are laid out."""

    PCM16 = 0
    FLOAT32 = 1
    WAV = 2


class AudioFormat(IntEnum):
    """Container of synthesized audio (:class:`~runanywhere.results.Audio`)."""

    PCM = 0
    WAV = 1


@dataclass(frozen=True)
class AudioFormatSpec:
    """Sample layout of an :class:`AudioInput`."""

    encoding: AudioEncoding
    sample_rate: int
    channels: int = 1


@dataclass(frozen=True)
class AudioInput:
    """Audio handed to ``stt``, ``vad`` or ``diarization``, in any supported layout.

    Build one with :meth:`pcm16`, :meth:`float32`, :meth:`wav` or :meth:`file` rather than
    the constructor.
    """

    data: bytes
    format: AudioFormatSpec

    @classmethod
    def pcm16(cls, data: bytes, sample_rate: int = STT_SAMPLE_RATE, channels: int = 1) -> "AudioInput":
        """Wrap signed 16-bit little-endian PCM bytes."""
        return cls(bytes(data), AudioFormatSpec(AudioEncoding.PCM16, sample_rate, channels))

    @classmethod
    def float32(cls, samples: "np.ndarray | List[float]", sample_rate: int) -> "AudioInput":
        """Wrap float samples in [-1, 1]."""
        arr = np.asarray(samples, dtype=np.float32)
        return cls(arr.tobytes(), AudioFormatSpec(AudioEncoding.FLOAT32, sample_rate, 1))

    @classmethod
    def wav(cls, data: bytes) -> "AudioInput":
        """Wrap the bytes of a 16-bit RIFF/WAVE file."""
        return cls(bytes(data), AudioFormatSpec(AudioEncoding.WAV, 0, 1))

    @classmethod
    def file(cls, path: str) -> "AudioInput":
        """Read a 16-bit WAV file from disk.

        Raises:
            SDKException: the file cannot be read.
        """
        try:
            with open(path, "rb") as handle:
                return cls.wav(handle.read())
        except OSError as exc:
            raise SDKException.storage_error(f"could not read audio file {path}: {exc}") from exc

    def samples(self, sample_rate: int = STT_SAMPLE_RATE) -> np.ndarray:
        """Decode to mono float32 samples resampled to ``sample_rate``."""
        if self.format.encoding == AudioEncoding.WAV:
            rate, samples = decode_wav(self.data)
        elif self.format.encoding == AudioEncoding.FLOAT32:
            rate = self.format.sample_rate
            samples = np.frombuffer(self.data, dtype=np.float32)
        else:
            rate = self.format.sample_rate
            raw = np.frombuffer(self.data, dtype="<i2").astype(np.float32) / 32768.0
            samples = raw
        if self.format.encoding != AudioEncoding.WAV and self.format.channels > 1:
            samples = samples.reshape(-1, self.format.channels).mean(axis=1)
        if not rate:
            raise SDKException.invalid_input("audio input has no sample rate")
        samples = np.asarray(samples, dtype=np.float32)
        if rate > sample_rate:
            samples = downsample(samples, rate, sample_rate)
        elif rate < sample_rate and samples.size:
            # audio.downsample only decimates, so interpolate when the input is slower than
            # the target — passing it through at the wrong rate would garble transcription.
            count = int(round(samples.size * sample_rate / rate))
            samples = np.interp(
                np.linspace(0.0, samples.size - 1, count),
                np.arange(samples.size),
                samples,
            ).astype(np.float32)
        return np.ascontiguousarray(samples, dtype=np.float32)

    def to_pcm16(self, sample_rate: int = STT_SAMPLE_RATE) -> bytes:
        """Encode to signed 16-bit mono PCM bytes at ``sample_rate``."""
        return pcm16_bytes(self.samples(sample_rate))

    def duration_ms(self, sample_rate: int = STT_SAMPLE_RATE) -> int:
        """Length of the audio in milliseconds."""
        return int(round(len(self.samples(sample_rate)) / sample_rate * 1000))


def _bmp24(data: bytes, width: int, height: int) -> bytes:
    """Encode packed RGB bytes as a 24-bit BMP (the container the engines' image loader reads)."""
    if width <= 0 or height <= 0:
        raise SDKException.invalid_input("image width and height must be positive")
    if len(data) != width * height * 3:
        raise SDKException.invalid_input(
            f"raw RGB data is {len(data)} bytes, expected {width * height * 3}"
        )
    pad = b"\x00" * ((-(width * 3)) % 4)  # BMP scanlines are 4-byte aligned
    rgb = np.frombuffer(data, dtype=np.uint8).reshape(height, width, 3)
    bgr = rgb[::-1, :, ::-1]  # bottom-up scanlines, BGR channel order
    pixels = b"".join(row.tobytes() + pad for row in np.ascontiguousarray(bgr))
    size = 54 + len(pixels)
    header = struct.pack("<2sIHHI", b"BM", size, 0, 0, 54)
    info = struct.pack("<IiiHHIIiiII", 40, width, height, 1, 24, 0, len(pixels), 0, 0, 0, 0)
    return header + info + pixels


@dataclass
class ImageInput:
    """Image handed to ``vlm`` or ``segmentation``.

    Build one with :meth:`file`, :meth:`bytes` or :meth:`raw_rgb`.
    ``segmentation`` needs decoded pixels — use :meth:`raw_rgb` (there is no image
    decoder in commons). ``vlm`` accepts a path or encoded bytes via :meth:`resolve_path`.
    """

    path: Optional[str] = None
    data: Optional[bytes] = None
    width: int = 0
    height: int = 0
    # Packed RGB8 pixels when built via :meth:`raw_rgb` (``rac_segmentation_image_t``).
    rgb: Optional[bytes] = None
    _temp: bool = False

    @classmethod
    def file(cls, path: str) -> "ImageInput":
        """Reference an image already on disk (JPEG/PNG/BMP)."""
        if not os.path.isfile(path):
            raise SDKException.invalid_input(f"image file not found: {path}")
        return cls(path=path)

    @classmethod
    def bytes(cls, data: bytes) -> "ImageInput":
        """Wrap encoded image bytes (JPEG/PNG/BMP)."""
        return cls(data=bytes(data))

    @classmethod
    def raw_rgb(cls, data: bytes, width: int, height: int) -> "ImageInput":
        """Wrap packed 8-bit RGB pixels, row-major from the top-left."""
        raw = bytes(data)
        return cls(
            data=_bmp24(raw, width, height),
            width=width,
            height=height,
            rgb=raw,
        )

    def resolve_path(self) -> str:
        """Return a filesystem path for this image, materializing bytes into a temp file."""
        if self.path is not None:
            return self.path
        if self.data is None:
            raise SDKException.invalid_input("image input carries neither a path nor bytes")
        fd, path = tempfile.mkstemp(suffix=".img")
        with os.fdopen(fd, "wb") as handle:
            handle.write(self.data)
        self.path = path
        self._temp = True
        return path

    def release(self) -> None:
        """Delete the temp file created by :meth:`resolve_path` (no-op for disk images)."""
        if self._temp and self.path is not None:
            try:
                os.unlink(self.path)
            except OSError:
                pass
            self.path = None
            self._temp = False


class Role(IntEnum):
    """Author of a :class:`ChatMessage`."""

    SYSTEM = 0
    USER = 1
    ASSISTANT = 2
    TOOL = 3


@dataclass
class ChatMessage:
    """One turn of a conversation passed to ``llm.generate``."""

    role: Role
    content: str
    tool_call_id: Optional[str] = None


@dataclass(frozen=True)
class ModelRef:
    """A model id, plus the voice to use when the model is a TTS voice bank."""

    id: str
    voice: Optional[str] = None


class ModelCategory(IntEnum):
    """What a model does. Values mirror ``rac_model_category_t`` (the C ABI, not proto wire)."""

    LANGUAGE = 0
    SPEECH_RECOGNITION = 1
    SPEECH_SYNTHESIS = 2
    VISION = 3
    IMAGE_GENERATION = 4
    MULTIMODAL = 5
    AUDIO = 6
    EMBEDDING = 7
    VOICE_ACTIVITY_DETECTION = 8
    SPEAKER_DIARIZATION = 9
    SEMANTIC_SEGMENTATION = 10
    UNKNOWN = 99


class InferenceFramework(IntEnum):
    """Engine that executes a model. Values mirror ``rac_inference_framework_t``."""

    ONNX = 0
    LLAMACPP = 1
    FOUNDATION_MODELS = 2
    SYSTEM_TTS = 3
    FLUID_AUDIO = 4
    BUILTIN = 5
    NONE = 6
    MLX = 7
    COREML = 8
    SHERPA = 12
    QHEXRT = 13
    UNKNOWN = 99


@dataclass
class ModelRegistration:
    """A model to add to the registry: one url, an archive, or a set of files.

    Exactly one of ``url``, ``files`` or ``path`` describes where the weights come from.
    """

    id: str
    category: ModelCategory
    url: Optional[str] = None
    files: List[str] = field(default_factory=list)
    archive: bool = False
    path: Optional[str] = None
    primary: Optional[str] = None
    framework: Optional[InferenceFramework] = None
    label: Optional[str] = None
    size_bytes: int = 0


@dataclass
class ModelFilter:
    """Narrows :meth:`runanywhere.models.list` to a category and/or download state."""

    category: Optional[ModelCategory] = None
    downloaded: Optional[bool] = None


@dataclass
class ToolDefinition:
    """A tool the model may call, described by its JSON-Schema parameters."""

    name: str
    parameters: JsonSchema = field(default_factory=lambda: {"type": "object"})
    description: Optional[str] = None


@dataclass
class RagDocument:
    """A document to ingest into a :class:`~runanywhere.api.rag.RagSession`."""

    text: str
    metadata: Optional[Dict[str, str]] = None
    id: str = ""

    @classmethod
    def file(cls, path: str) -> "RagDocument":
        """Read a UTF-8 text file, tagging the document with its path.

        Raises:
            SDKException: the file cannot be read.
        """
        try:
            with open(path, "r", encoding="utf-8") as handle:
                text = handle.read()
        except OSError as exc:
            raise SDKException.storage_error(f"could not read document {path}: {exc}") from exc
        return cls(text=text, metadata={"source": path}, id=os.path.basename(path))

File: sdk/runanywhere-kotlin/src/main/kotlin/com/runanywhere/sdk/public/api/Inputs.kt
/*
 * Copyright 2026 RunAnywhere SDK
 * SPDX-License-Identifier: Apache-2.0
 *
 * Public v3 surface: the input envelopes every modality accepts.
 */

package com.runanywhere.sdk.public.api

import android.graphics.Bitmap
import com.runanywhere.sdk.foundation.errors.SDKException
import com.runanywhere.sdk.native.bridge.RunAnywhereBridge
import java.io.File
import java.nio.ByteBuffer
import java.nio.ByteOrder

/** Sample layout of a raw audio buffer. */
public enum class AudioEncoding {
    /** Signed 16-bit little-endian samples. */
    PCM16,

    /** 32-bit float samples in `[-1.0, 1.0]`. */
    FLOAT32,

    /** Self-describing container such as WAV. */
    CONTAINER,
}

/** How to interpret the bytes of an [AudioInput]. */
public data class AudioFormatSpec(
    val encoding: AudioEncoding,
    val sampleRate: Int = DEFAULT_SAMPLE_RATE,
    val channels: Int = 1,
) {
    public companion object {
        public const val DEFAULT_SAMPLE_RATE: Int = 16_000
    }
}

/**
 * Audio handed to the SDK, together with the format needed to read it.
 *
 * Build one with [pcm16], [float32], [wav], or [file] rather than the
 * constructor, so the format spec always matches the bytes.
 */
public class AudioInput internal constructor(
    public val bytes: ByteArray,
    public val format: AudioFormatSpec,
) {
    /** Bytes normalised to the float32 layout every commons audio path expects. */
    internal fun normalizedBytes(): ByteArray =
        when (format.encoding) {
            AudioEncoding.PCM16 -> pcm16ToFloat32Bytes(bytes)
            AudioEncoding.FLOAT32, AudioEncoding.CONTAINER -> bytes
        }

    /**
     * Bytes as signed 16-bit little-endian PCM — the layout the sherpa STT
     * engine expects (it reinterprets the buffer as int16 and converts to float
     * itself; see the iOS bridge, which sends unconverted PCM16). Sending
     * float32 here would be misread as int16 and transcribed as noise.
     */
    internal fun pcm16Bytes(): ByteArray =
        when (format.encoding) {
            AudioEncoding.PCM16, AudioEncoding.CONTAINER -> bytes
            AudioEncoding.FLOAT32 -> float32ToPcm16Bytes(bytes)
        }

    public companion object {
        /** Wrap signed 16-bit little-endian microphone samples. */
        public fun pcm16(
            bytes: ByteArray,
            sampleRate: Int = AudioFormatSpec.DEFAULT_SAMPLE_RATE,
            channels: Int = 1,
        ): AudioInput = AudioInput(bytes, AudioFormatSpec(AudioEncoding.PCM16, sampleRate, channels))

        /** Wrap float samples already normalised to `[-1.0, 1.0]`. */
        public fun float32(
            samples: FloatArray,
            sampleRate: Int = AudioFormatSpec.DEFAULT_SAMPLE_RATE,
        ): AudioInput {
            val buffer = ByteBuffer.allocate(samples.size * Float.SIZE_BYTES).order(ByteOrder.LITTLE_ENDIAN)
            samples.forEach(buffer::putFloat)
            return AudioInput(buffer.array(), AudioFormatSpec(AudioEncoding.FLOAT32, sampleRate))
        }

        /** Wrap float32 bytes that are already little-endian. */
        public fun float32Bytes(
            bytes: ByteArray,
            sampleRate: Int = AudioFormatSpec.DEFAULT_SAMPLE_RATE,
        ): AudioInput = AudioInput(bytes, AudioFormatSpec(AudioEncoding.FLOAT32, sampleRate))

        /** Wrap a complete WAV container. */
        public fun wav(bytes: ByteArray): AudioInput =
            AudioInput(bytes, AudioFormatSpec(AudioEncoding.CONTAINER))

        /**
         * Read an audio file from disk.
         *
         * @throws SDKException when the file is missing or unreadable.
         */
        public fun file(path: String): AudioInput {
            val data = readFileOrThrow(path, "audio")
            return AudioInput(data, AudioFormatSpec(AudioEncoding.CONTAINER))
        }
    }
}

/** Raw pixel layout of an [ImageInput]. */
public enum class ImagePixelLayout {
    RGB8,
    RGBA8,
}

/**
 * An image handed to the SDK.
 *
 * Build one with [file], [bytes], [rawRgb], [rawRgba], or [bitmap].
 */
public class ImageInput private constructor(
    internal val encoded: ByteArray?,
    internal val raw: ByteArray?,
    internal val width: Int,
    internal val height: Int,
    internal val layout: ImagePixelLayout?,
    internal val filePath: String?,
) {
    public companion object {
        /** Reference an image on disk without reading it into memory. */
        public fun file(path: String): ImageInput =
            ImageInput(null, null, 0, 0, null, path)

        /** Wrap encoded JPEG, PNG, or WebP bytes. */
        public fun bytes(data: ByteArray): ImageInput =
            ImageInput(data, null, 0, 0, null, null)

        /** Wrap tightly packed 3-byte-per-pixel RGB data. */
        public fun rawRgb(data: ByteArray, width: Int, height: Int): ImageInput =
            ImageInput(null, data, width, height, ImagePixelLayout.RGB8, null)

        /** Wrap tightly packed 4-byte-per-pixel RGBA data. */
        public fun rawRgba(data: ByteArray, width: Int, height: Int): ImageInput =
            ImageInput(null, data, width, height, ImagePixelLayout.RGBA8, null)

        /** Wrap an Android [Bitmap], packing it as RGBA. */
        public fun bitmap(bitmap: Bitmap): ImageInput {
            val width = bitmap.width
            val height = bitmap.height
            val pixels = IntArray(width * height)
            bitmap.getPixels(pixels, 0, width, 0, 0, width, height)
            val rgba = ByteArray(pixels.size * 4)
            pixels.forEachIndexed { index, pixel ->
                val offset = index * 4
                rgba[offset] = ((pixel shr 16) and 0xFF).toByte()
                rgba[offset + 1] = ((pixel shr 8) and 0xFF).toByte()
                rgba[offset + 2] = (pixel and 0xFF).toByte()
                rgba[offset + 3] = ((pixel shr 24) and 0xFF).toByte()
            }
            return rawRgba(rgba, width, height)
        }
    }
}

/** Who authored a message in a chat transcript. */
public enum class ChatRole {
    SYSTEM,
    USER,
    ASSISTANT,
    TOOL,
}

/** One turn of a chat transcript. */
public data class ChatMessage(
    val role: ChatRole,
    val content: String,
    val toolCallId: String? = null,
)

/** Points a session or generation at a registered model, optionally at one voice inside it. */
public data class ModelRef(
    val id: String,
    val voice: String? = null,
)

/** A document to index into a RAG session. */
public class RagDocument private constructor(
    internal val text: String,
    internal val metadata: Map<String, String>,
    internal val sourceUri: String?,
) {
    public constructor(text: String, metadata: Map<String, String>? = null) :
        this(text, metadata.orEmpty(), null)

    public companion object {
        /**
         * Read a UTF-8 text document from disk.
         *
         * @throws SDKException when the file is missing or unreadable.
         */
        public fun file(path: String): RagDocument {
            val data = readFileOrThrow(path, "document")
            return RagDocument(data.decodeToString(), emptyMap(), path)
        }
    }
}

private fun readFileOrThrow(path: String, kind: String): ByteArray {
    val handle = File(path)
    if (!handle.isFile) {
        throw SDKException.invalidArgument("No $kind file at $path")
    }
    return runCatching { handle.readBytes() }.getOrElse { error ->
        throw SDKException.invalidArgument("Could not read $kind file at $path", error)
    }
}

private fun pcm16ToFloat32Bytes(pcm16: ByteArray): ByteArray {
    val samples =
        RunAnywhereBridge.racAudioPcm16ToFloat32(pcm16)
            ?: throw SDKException.invalidArgument("rac_audio_pcm16_to_float32 failed")
    if (samples.isEmpty()) return ByteArray(0)
    val output = ByteBuffer.allocate(samples.size * Float.SIZE_BYTES).order(ByteOrder.LITTLE_ENDIAN)
    for (sample in samples) {
        output.putFloat(sample)
    }
    return output.array()
}

private fun float32ToPcm16Bytes(float32: ByteArray): ByteArray {
    val sampleCount = float32.size / Float.SIZE_BYTES
    if (sampleCount == 0) return ByteArray(0)
    val input = ByteBuffer.wrap(float32).order(ByteOrder.LITTLE_ENDIAN)
    val samples = FloatArray(sampleCount) { input.float }
    return RunAnywhereBridge.racAudioFloat32ToPcm16(samples)
        ?: throw SDKException.invalidArgument("rac_audio_float32_to_pcm16 failed")
}
